### ISSUE TITLE
Code Demo

### ISSUE THREAD

**Issue by harrylee999** [woodenchild95/FL-Simulator#17]
Heey， can u provide the code demo of FedSMOO with CIFAR100, I run 
`python train.py --non-iid --dataset CIFAR-100 --model ResNet18 --split-rule Dirichlet --split-coef 0.1 --active-ratio 0.1 --total-client 100 --local-epoch 2 --batchsize 20 --lr-decay 0.9995 --comm-rounds 800 --rho 0.1 --method FedSMOO`

The performance of FedSMOO even not better than FedAvg.

**Comment by woodenchild95** [comment]
@harrylee999 see u~ Thank you very much for continuing to follow our work. Regarding the experimental section, you might consider using CIFAR-10 training setups with batch size = 50, local epoch = 5, learning rate decay = 0.998, and rho = 0.1.

Last year, I received feedback from some researchers, but I noticed that the optimal selection varies across different groups. Some successfully reproduce my results, while others have to re-tune the hyperparameters and then achieving results consistent with the paper. I'm not sure if this discrepancy is due to differences in computing environments, as some of the hyperparameters they eventually found to work did not reproduce the expected results on my local device.

If your computational resources allow, I would recommend exploring the hyperparameter space according to your own experimental setup.

**Comment by harrylee999** [comment]
Thanks for your quick response.

My problem is my method can get acc about 50% at the setting with batch size = 20, local epoch = 2 as following paper setting, and FedSMOO always failed on this setting (I change random seed it still not work).

Acutually, I try batch size = 50, local epoch = 5, learning rate decay = 0.998, and rho = 0.1, FedSMOO can get acc about 49% at last. But my method will get acc 47% at this setting.

So Idk how to set the basic hyperparameters for fair comparsion. Do u have any advice?

By the way, how do u compute the standard deviation?  The paper said each result is calculated by 2 times, which seed you used？

**Comment by woodenchild95** [comment]
@harrylee999 Fair comparison is a tricky concept. Some people believe that it's reasonable for different algorithms to search for different hyperparameters to achieve optimal accuracy, while others think it's wrong, and that hyperparameters should be fixed across algorithms. As you mentioned, the performance difference across different batch sizes, I think you could create a grid search plot to show the performance of different algorithms at different batch sizes. This is also a way to demonstrate the stability of algorithms, by performing "hyperparameter sensitivity experiments" across a broad range.

The standard deviation can be calculated by changing the random seed, which generally reflects the performance fluctuation under a fixed set of hyperparameters. The set of hyperparameters mentioned above was just one of the responses I received before, but in reality, there are many such "magic numbers." I think that the differences among results stem from variations in computational libraries and environmental discrepancies.

The fairness comparison I recommend is to define a reasonable range for all hyperparameters, such as batch_size = [10, 20, 50, 100], lr_decay = [0.998, 0.9998, 1], etc., to include combinations where all methods can perform well. Then, systematically explore these combinations and search for the best performance for each method. This approach can help mitigate experimental biases caused by unknown factors.

Wishing you success in your experiments! 😊

**Comment by harrylee999** [comment]
@woodenchild95 Thanks for your worthful advice! I get the result I want.
In addition, another work [FedLESAM](https://arxiv.org/pdf/2405.18890) in ICML2024, which directly took the experiment results of FedSMOO's in their paper. In my opinion, this is an unfair comparsion because different device produces different result. I also implented FedLESAM and its variant in this repo which worse than FedSMOO in most settings.

**Comment by woodenchild95** [comment]
@harrylee999 Many times, the motivation behind a method in a paper is crucial. Be sure to clearly articulate the motivation, then quantify the problem before solving it. Their paper is a great example of this. The experimental results section can sometimes be quite subtle, as each method has its own applicable scope. As long as you conduct a "relatively fair comparison" in your experiments, that's enough. Personally, I really dislike the review style that solely focuses on SOTA results. I believe that the motivation behind doing something is far more important than the results—it inspires more people to think together.

Wishing you success in your research. Surpass them and get accepted to NIPS 2025! 💪🔥If you have any related questions, feel free to contact me.
